src/drivers/uart/src/i8250.rs

```diff
@@ -95,43 +99,78 @@ impl<'a> Driver for I8250<'a> {
     fn shutdown(&mut self) {}
 }
 
-// // TODO: bitfields
-// register_bitfields! {
-//     u8,
-//     // Data register
-//     D [
-//         DATA OFFSET(0) NUMBITS(8) []
-//     ],
-//     IE [
-//         RX OFFSET(0) NUMBITS(1) [],
-//         TX OFFSET(1) NUMBITS(1) [],
-//         Error OFFSET(2) NUMBITS(1) [],
-//         StatusChange OFFSET(3) NUMBITS(1) []
-//     ],
-//     FC [
-//         DATA OFFSET(0) NUMBITS(8) []
-//     ],
-//     LC [
-//         WLEN OFFSET(0) NUMBITS(2) [
-//             WLEN_5 = 0,
-//             WLEN_6 = 1,
-//             WLEN_7 = 2,
-//             WLEN_8 = 3
-//         ],
-//         StopBits OFFSET(3) NUMBITS(1) [],
-//         ParityEnable OFFSET(4) NUMBITS(1) [],
-//         EvenParity OFFSET(5) NUMBITS (1) [],
-//         StickParity OFFSET(6) NUMBITS (1) [],
-//         DivisorLatchAccessBit OFFSET(7) NUMBITS (1) [
-//             Normal = 0,
-//             BaudRate = 1
-//         ]
-//     ],
-//     MC [
-//         DATA OFFSET(0) NUMBITS(8) []
-//     ],
-//     LS [
-//         IF OFFSET(0) NUMBITS(1) [],
-//         OE OFFSET(1) NUMBITS(1) []
-//     ]
-// }
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /*
+     * UART pushing into a vec
+     */
+    extern crate heapless;
+    use heapless::Vec;
+    pub struct MockPort<'a> {
+        ldata: &'a mut Vec<u8, heapless::consts::U8>,
+    }
+
+    impl<'a> MockPort<'a> {
+        pub fn new(v: &'a mut Vec<u8, heapless::consts::U8>) -> MockPort {
+            MockPort { ldata: v }
+        }
+    }
+
+    impl<'a> Driver for MockPort<'a> {
+        fn init(&mut self) -> Result<()> {
+            Ok(())
+        }
+
+        fn pread(&self, data: &mut [u8], offset: usize) -> Result<usize> {
+            if self.ldata.len() <= offset {
+                return EOF;
+            }
+            data[0] = self.ldata[offset];
+            return Ok(0);
+        }
+
+        fn pwrite(&mut self, data: &[u8], offset: usize) -> Result<usize> {
+            while self.ldata.len() < offset + data.len() {
+                self.ldata.push(0).unwrap();
+            }
+
+            for (i, &c) in data.iter().enumerate() {
+                self.ldata[offset + i] = c;
+            }
+
+            Ok(data.len())
+        }
+
+        fn shutdown(&mut self) {}
+    }
+
+    const FCR: usize = 0x02; // FIFO Control Register                0b0010 W
+    const LCR: usize = 0x03; // Line Control Register                0b0011 RW
+                             // const DLL: usize = 0x00; // Divisor Latch Low Byte               0      RW
+                             // const DLH: usize = 0x01; // Divisor Latch High Byte              0x0001 RW
```

Comment:
Remove this commented code if it's not used.

Reply:
Done